Use time.Time instead of uint64 for genesis time

prestonvanloon opened the pull request:
**What type of PR is this?**

Other

**What does this PR do? Why is it needed?**

Using `uint64` AND `time.Time` has some pains. I've changed Prysm to use time.Time whenever possible, except to comply with spec. 

**Which issues(s) does this PR fix?**

**Other notes for review**

This is a pre-requisite PR for dynamic slot times.

**Acknowledgements**

- [x] I have read [CONTRIBUTING.md](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md).
- [x] I have included a uniquely named [changelog fragment file](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md#maintaining-changelogmd).
- [x] I have added a description to this PR with sufficient context for reviewers to understand this PR.

prestonvanloon committed:
Convert genesis times from seconds to time.Time

prestonvanloon committed:
Fixing failed forkchoice tests in a new commit so it doesn't get worse

Fixing failed spectest tests in a new commit so it doesn't get worse

Fixing forkchoice tests, then spectests

prestonvanloon committed:
Fixing forkchoice tests, then spectests. Now asking for help...

potuz committed:
Fix TestForkChoice_GetProposerHead

prestonvanloon committed:
Fix broken build

prestonvanloon committed:
Resolve TODO(preston) items

prestonvanloon committed:
Changelog fragment

prestonvanloon committed:
Resolve TODO(preston) items again

prestonvanloon committed:
Resolve lint issues

terencechain commented on beacon-chain/blockchain/forkchoice_update_execution.go:
```diff
@@ -132,16 +132,16 @@ func (s *Service) shouldOverrideFCU(newHeadRoot [32]byte, proposingSlot primitiv
-			}).Infof("Attempted late block reorg aborted due to attestations at %d seconds",
+				"root":             fmt.Sprintf("%#x", newHeadRoot),
+				"weight":           headWeight,
+				"since slot start": sss,
```
should this be `sinceSlotStart`, we dont typically put spaces as field do we?

prestonvanloon replied:
Yes, I agree. Thanks!

prestonvanloon committed:
Use consistant field names for sinceSlotStart (no spaces)

nalepae commented on beacon-chain/blockchain/forkchoice_update_execution.go:
```diff
@@ -132,16 +132,16 @@ func (s *Service) shouldOverrideFCU(newHeadRoot [32]byte, proposingSlot primitiv
+				"root":           fmt.Sprintf("%#x", newHeadRoot),
+				"weight":         headWeight,
+				"sinceSlotStart": sss,
+			}).Infof("Attempted late block reorg aborted due to attestations after %s",
```
Comment 1: There is no time unit any more for the time in the info message.
Comment 2: Instead of using `Infof`, better to add a new `processAttestationThreshold` in fields and use `Info`?

prestonvanloon replied:
> Comment 1: There is no time unit any more for the time in the info message.

[time.Duration.String()](https://pkg.go.dev/time#Duration.String) will print the time units.

> Comment 2: Instead of using Infof, better to add a new processAttestationThreshold in fields and use Info

Yeah, that is fair, i'll fix

nalepae commented on testing/endtoend/components/eth1/miner.go:
```diff
@@ -89,7 +89,7 @@ func (m *Miner) initAttempt(ctx context.Context, attempt int) (*os.File, error)
 	gethJsonPath := path.Join(path.Dir(binaryPath), "genesis.json")
 	gen := interop.GethTestnetGenesis(e2e.TestParams.Eth1GenesisTime, params.BeaconConfig())
-	log.Infof("eth1 miner genesis timestamp=%d", e2e.TestParams.Eth1GenesisTime)
+	log.Infof("eth1 miner genesis timestamp=%s", e2e.TestParams.Eth1GenesisTime)
```
Can we use a logrus field instead `Infof`?

prestonvanloon committed:
Manu's feedback

james-prysm commented on beacon-chain/blockchain/options.go:
```diff
@@ -246,7 +246,7 @@ func WithSlasherEnabled(enabled bool) Option {
 func WithGenesisTime(genesisTime time.Time) Option {
 	return func(s *Service) error {
-		s.genesisTime = genesisTime
+		s.genesisTime = genesisTime.Truncate(time.Second) // Genesis time has a precision of 1 second.
```
does it round up or down ( does that change how it used to work?)

prestonvanloon replied:
It rounds down. This is mostly a convenience for test setup. Spec requires "genesis time" to be some "seconds since unix epoch" which will not have milliseconds, nanoseconds, etc.

james-prysm commented on testing/util/fulu_block.go:
```diff
@@ -109,7 +109,7 @@ func GenerateFullBlockFulu(bState state.BeaconState, privs []bls.SecretKey, conf
 	}
 
-	timestamp, err := slots.ToTime(bState.GenesisTime(), slot)
+	timestamp, err := slots.SlotTime(bState.GenesisTime(), slot)
```
what about slots.Time? slot.slottime sounds like it's stuttering

prestonvanloon replied:
slots.StartTime seems appropriate here. The method returns the time in which the slot begins, so it makes sense to me. Thanks

james-prysm commented on beacon-chain/blockchain/process_attestation_helpers.go:
```diff
@@ -139,8 +139,8 @@ func (s *Service) getAttPreState(ctx context.Context, c *ethpb.Checkpoint) (stat
-func verifyAttTargetEpoch(_ context.Context, genesisTime, nowTime uint64, c *ethpb.Checkpoint) error {
-	currentSlot := primitives.Slot((nowTime - genesisTime) / params.BeaconConfig().SecondsPerSlot)
+func verifyAttTargetEpoch(_ context.Context, genesis, now time.Time, c *ethpb.Checkpoint) error {
+	currentSlot := slots.SlotAt(genesis, now)
```
what about `slots.At(genesis, now)`

prestonvanloon replied:
Sure, that's fine with me. Thanks

james-prysm commented on beacon-chain/blockchain/receive_block_test.go:
```diff
@@ -348,8 +349,9 @@ func TestHandleCaches_EnablingLargeSize(t *testing.T) {
-	s.genesisTime = time.Now().Add(time.Duration(-1*int64(st)*int64(params.BeaconConfig().SecondsPerSlot)) * time.Second)
+	s.SetGenesisTime(time.Now().Add(time.Duration(-1*int64(st)*int64(params.BeaconConfig().SecondsPerSlot)) * time.Second))
 
+	helpers.ClearCache()
```
why did we need to add this?

prestonvanloon replied:
Not sure, i'll remove it

prestonvanloon replied:
Tests fail without it

prestonvanloon replied:
I think conflicts with the global cache now that it is using real times. I'll leave it

james-prysm commented on beacon-chain/core/helpers/attestation.go:
```diff
@@ -10,7 +10,6 @@ import (
 	"github.com/OffchainLabs/prysm/v6/consensus-types/primitives"
 	"github.com/OffchainLabs/prysm/v6/crypto/hash"
 	ethpb "github.com/OffchainLabs/prysm/v6/proto/prysm/v1alpha1"
-	prysmTime "github.com/OffchainLabs/prysm/v6/time"
```
do we not need to use this anymore?

prestonvanloon replied:
I want to remove it. It's a wrapper around stdlib time and no longer serves any purpose imo

james-prysm commented on cmd/prysmctl/testnet/generate_genesis.go:
```diff
@@ -280,7 +281,7 @@ func generateGenesis(ctx context.Context) (state.BeaconState, error) {
 		}
 	} else {
-		gen = interop.GethTestnetGenesis(f.GenesisTime, params.BeaconConfig())
+		gen = interop.GethTestnetGenesis(time.Unix(int64(f.GenesisTime), 0), params.BeaconConfig())
```
i see the pattern time.Unix(int64(f.GenesisTime), 0) quite a bit, is it worth having some helper function for it?

prestonvanloon replied:
All of this is going to change in the slot time schedule stuff so let's defer those convenience methods to the follow up PRs to accommodate a dynamic slot time.

james-prysm commented on testing/util/bellatrix_state.go:
```diff
@@ -27,7 +28,7 @@ func DeterministicGenesisStateBellatrix(t testing.TB, numValidators uint64) (sta
 		t.Fatal(errors.Wrapf(err, "failed to get eth1data for %d deposits", numValidators))
 	}
-	beaconState, err := genesisBeaconStateBellatrix(context.Background(), deposits, uint64(0), eth1Data)
+	beaconState, err := genesisBeaconStateBellatrix(context.Background(), deposits, time.Unix(0, 0), eth1Data)
```
you can change context.Background() to t.Context() i think

prestonvanloon committed:
Renamed StartTime -> UnsafeStartTime, marked as deprecated because it doesn't handle overflow scenarios.
Renamed SlotTime -> StartTime
Renamed SlotAt -> At
Handled the error in cases where StartTime was used.

@james-prysm feedback

prestonvanloon committed:
Revert beacon-chain/blockchain/receive_block_test.go from 1b7844de

prestonvanloon committed:
Fixing issues after rebase

potuz commented on beacon-chain/blockchain/process_block_test.go:
```diff
@@ -2319,17 +2321,24 @@ func TestFillMissingBlockPayloadId_PrepareAllPayloads(t *testing.T) {
-	s.SetGenesisTime(newTime)
-	s.cfg.ForkChoiceStore.SetGenesisTime(uint64(newTime.Unix()))
+func driftGenesisTime(s *Service, slot primitives.Slot, delay time.Duration) {
+	now := time.Now()
+	slotStart, err := slots.StartTime(now, slot)
+	if err != nil {
+		panic(err) // lint:nopanic -- This is test code and should never overflow.
+	}
+	offset := now.Sub(slotStart).Abs() + delay
+	genesis := now.Add(-offset)
```
You can avoid the panic with
```suggestion
  	now := time.Now()
  	slotDuration := time.Duration(slot) * time.Duration(params.BeaconConfig().SecondsPerSlot) * time.Second
  	genesis := now.Add(-slotDuration - delay)
```

potuz commented on beacon-chain/blockchain/process_block.go:
```diff
@@ -855,7 +858,10 @@ func (s *Service) areBlobsAvailable(ctx context.Context, root [fieldparams.RootL
-	nextSlot := slots.BeginsAt(block.Slot()+1, s.genesisTime)
+	nextSlot, err := slots.StartTime(s.genesisTime, block.Slot()+1)
+	if err != nil {
+		log.WithError(err).Error("unable to determine slot start time")
```
Should this return an error as in the check above in line 741?

prestonvanloon replied:
yes, i think that is a good idea

potuz reviewed (dismissed):
LGTM

prestonvanloon committed:
Accepted suggestions from @potuz

james-prysm commented on validator/client/validator_test.go:
```diff
@@ -294,6 +294,37 @@ func TestWaitForChainStart_ReceiveErrorFromStream(t *testing.T) {
 	assert.ErrorContains(t, want, err)
 }
 
+func TestCanonicalHeadSlot_FailedRPC(t *testing.T) {
```
we don't need this, this is being added back in

james-prysm commented on validator/client/validator.go:
```diff
@@ -138,10 +138,18 @@ func (v *validator) Done() {
 	return v.genesisTime
 }
 
+func (v *validator) CanonicalHeadSlot(ctx context.Context) (primitives.Slot, error) {
```
we don't need this function, it was removed in https://github.com/OffchainLabs/prysm/pull/15480

james-prysm reviewed (changes requested):
please remove the canonical head api calls that were added back in

prestonvanloon committed:
Remove CanonicalHeadSlot from merge conflicts
